apollo-router/src/registry/mod.rs

```diff
@@ -0,0 +1,175 @@
+use std::string::FromUtf8Error;
+
+use docker_credential::CredentialRetrievalError;
+use docker_credential::DockerCredential;
+use oci_client::Client as ociClient;
+use oci_client::Reference;
+use oci_client::errors::OciDistributionError;
+use oci_client::secrets::RegistryAuth;
+use thiserror::Error;
+
+/// Configuration for fetching an OCI Bundle
+/// This struct does not change on router reloads - they are all sourced from CLI options.
+#[derive(Debug, Clone)]
+pub struct OCIConfig {
+    /// The Apollo key: `<YOUR_GRAPH_API_KEY>`
+    pub apollo_key: String,
+
+    // Should this be a reference instead?
+    pub url: String,
+}
+
+pub(crate) struct OCIResult {
+    pub schema: String,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum Error {
+    #[error("OCI layer has not title")]
+    OCILayerMissingTitle,
+    #[error("Oci Distribution error: {0}")]
+    OCIDistributionError(OciDistributionError),
+    #[error("Oci Parsing error: {0}")]
+    OCIParseError(oci_client::ParseError),
+    #[error("Unable to parse layer: {0}")]
+    OCILayerParseError(FromUtf8Error),
+}
+
+impl From<oci_client::ParseError> for Error {
+    fn from(value: oci_client::ParseError) -> Self {
+        Error::OCIParseError(value)
+    }
+}
+
+impl From<OciDistributionError> for Error {
+    fn from(value: OciDistributionError) -> Self {
+        Error::OCIDistributionError(value)
+    }
+}
+
+impl From<FromUtf8Error> for Error {
+    fn from(value: FromUtf8Error) -> Self {
+        Error::OCILayerParseError(value)
+    }
+}
+
+fn build_auth(reference: &Reference, apollo_key: &String) -> RegistryAuth {
```

Comment:
Never use `&String`. If it is a ref, use `&str` For more details on why: https://users.rust-lang.org/t/string-str-string-str/100611/3